src/main/java/org/wildfly/security/sasl/util/AuthenticationTimeoutSaslServerFactory.java

```diff
@@ -0,0 +1,126 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.sasl.SaslServer;
+import javax.security.sasl.SaslServerFactory;
+import javax.security.sasl.SaslException;
+
+import org.wildfly.common.Assert;
+import org.wildfly.security.sasl.WildFlySasl;
+
+/**
+ * A {@link SaslServerFactory} which adds authentication timeout functionality to a delegate {@code SaslServerFactory}.
+ * <p>
+ * This {@link SaslServerFactory} must be outside of the {@link AuthenticationCompleteCallbackSaslServerFactory} in the
+ * chain of SASL server factories.
+ *
+ * @author <a href="mailto:fjuma@redhat.com">Farah Juma</a>
+ */
+public final class AuthenticationTimeoutSaslServerFactory extends AbstractDelegatingSaslServerFactory {
+
+    /**
+     * The default amount of time, in seconds, after which an authentication attempt should time out.
+     */
+    public static final long DEFAULT_TIMEOUT = 150;
+
+    private final ScheduledExecutorService scheduledExecutorService;
+
+    /**
+     * Construct a new instance.
+     *
+     * @param delegate the delegate {@code SaslServerFactory}
+     * @param scheduledExecutorService the scheduled executor to use to handle authentication timeout tasks
+     */
+    public AuthenticationTimeoutSaslServerFactory(final SaslServerFactory delegate, final ScheduledExecutorService scheduledExecutorService) {
+        super(delegate);
+        Assert.checkNotNullParam("scheduledExecutorService", scheduledExecutorService);
+        this.scheduledExecutorService = scheduledExecutorService;
+    }
+
+    @Override
+    public SaslServer createSaslServer(final String mechanism, final String protocol, final String serverName, final Map<String, ?> props, final CallbackHandler cbh) throws SaslException {
+        final long timeout;
+        if (props.containsKey(WildFlySasl.AUTHENTICATION_TIMEOUT)) {
+            timeout = Long.parseLong((String) props.get(WildFlySasl.AUTHENTICATION_TIMEOUT));
+        } else {
+            timeout = DEFAULT_TIMEOUT;
+        }
+
+        final SaslServer delegateSaslServer = delegate.createSaslServer(mechanism, protocol, serverName, props, cbh);
+        return (delegateSaslServer == null) ? null : new DelegatingTimeoutSaslServer(delegateSaslServer, scheduledExecutorService, timeout);
+    }
+
+    private static class DelegatingTimeoutSaslServer extends AbstractDelegatingSaslServer {
+        private final AtomicBoolean complete = new AtomicBoolean();
+        private volatile ScheduledFuture<Void> timeoutTask;
+
+        DelegatingTimeoutSaslServer(final SaslServer delegate, final ScheduledExecutorService scheduledExecutorService, final long timeout) {
+            super(delegate);
+
+            // Schedule a task to terminate the authentication attempt if it takes too long
+            timeoutTask = scheduledExecutorService.schedule(() -> {
+                if (! (isComplete() && complete.compareAndSet(false, true))) {
```

Comment:
I think that because the executor service could fire at any time, there has to be a concurrency guard of some sort.  The "complete" flag doesn't suffice because it's not checked until late in evaluateResponse.  The simplest solution is to use synchronized to protect all access to the delegate SaslServer plus the "complete" flag and "timeoutTask" field, which means that the atomic boolean and volatile field can be simplified to plain fields.